0018-logging.adoc

```diff
@@ -0,0 +1,94 @@
+* warn: should be used when something bad happened, but the application still has the chance to heal itself or the issue can wait a day or two to be fixed.
+* info: should be used to document state changes in the application or some entity within the application.
+* debug: should be used to log any information that helps us identify what went wrong on DEBUG level.
+* trace: log all information that helps us to trace the flow of the application.
+
+=== Useful Context ===
+
+Useful context is not that easy to solve due to the async nature of cnd.
+Typically one can follow the sequence of log messages to get more context than is available within the scope of the log statement.
+The solution to this is most likely to use a tracing library instead of logging, see below for more on tracing.
+
+Related to context is useful output in the log messages i.e., don't dump byte slices using "{:?}".
+
+Suggestions to remedy this include:
+* Implement custom `Debug` for any objects that we would like to print within logs.
+* Add a `Log` trait for any objects that we wish to log and call the method when logging i.e., `log::info("Got: {}", foo.log())`
+
+=== Controlling Log Levels ===
+
+Currently we can control the base log level within the config file.  More fine grained control requires code changes in `cnd/src/logging/initialize.rs`.
+If we stay with logging I would suggest that we set the cnd log level to `trace`.
+It may be useful to have a few simple scripts in tree that do typical useful grep commands on the logs (unnecessary if we go with tracing).
+
+=== Tracing ===
+
+As mentioned in the issue on GitHub (linked above), logging may not be the correct approach because of the asynchronous nature of cnd.
+The issue is that because logging is event driven and the logs are ordered so it is hard to follow, from reading the log, what events are linked to each other.
+Tracing may be the solution to this but adding tracing requires a non-trivial implementation effort as well as some time investment for all devs on the team to learn how to debug using the trace output and interact with the tracing system.
+On top of this, tracing is most useful for, as the name suggests, 'tracing' a running application.
+Currently, our pain points come from running a failing e2e test, looking at the logs, and not being able to debug the issue.
+If used like this, we will not be leveraging the full potential of tracing, but at least we would be setting the ground work for this style of debugging in the future.
+Having the additional context and structured data may add value and help with debugging.
+
+There are at least two async tracing libraries in the Rust ecosystem, the most popular is by far `tracing` by the tokio folks.
+
+
+== Guidelines/suggestions ==
+
+In summary here are some proposed guidelines to help improve our logs.
+
+I would like to suggest that we attempt to utilise the current logging system fully before deciding to switch to tracing.
```

Comment:
Without switching to tracing, I will they will literally be useless unless we add the swap id manually to every log statement as soon as there is more than one swap going on.

I would even go as far and say, before we haven't switched to tracing, we are not production ready.

Reply:
This is an interesting point.  Is the improvement of logging an effort to enable us to debug issues we face _currently_ or is it an effort to be production ready? - if it is to be production ready then I am in total agreeance that we should use tracing.  I assumed we needed a better way to debug issues we are currently facing and were looking for the best and cheapest way to achieve that.  After your response I'll add a section explicitly stating that assumption to the spike.

Reply:
Before you do that, enjoy your holidays :)